XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+| Title  | AsyncWrap API |
+|--------|---------------|
+| Author | @trevnorris   |
+| Status | DRAFT         |
+| Date   | 2016-04-05    |
+
+## Description
+
+Since its initial introduction with the `AsyncListener` API, `AsyncWrap` has
+slowly evolved to ensure a generalized API that would serve as a solid base for
+module authors who wished to add events to the life cycle of the event loop.
+The API presented below aims to accomplish this.
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous providers (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const providers = async_wrap.Providers;
+
+// init() is called during object construction. The object handle being
+// constructed can be accessed as "this". Because init() is called during
+// construction it is not safe to blindly inspect the constructed object at the
+// moment. Though we are currently investigating how to make this safe.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
+// time for requests (e.g. FSREQWRAP).
+function pre(id) { }
```

Comment:
> Not clear how that will work for the `destroy()` case where there is no handle

There's always a handle. Just not passed in `destroy()` to prevent accidental abort, but I'm currently working on fixing this so the instance can be safely accessed at any time.

> Such a type can evolve to include appropriate information such as the type, handle-specific state, etc...

All this information already lives in the native layer. It would be far more cost effective and less complicated to query that information.

> The other useful bit of information is the _invocation count_. i.e., in cases where pre()/post() can be called 0..n times, what will uniquely identify an invocation?

The invocation count can be easily enough tracked by the user by a combination of tracking each request id to each handle that created it, and then when `pre()` is called by the request. Those are mechanics that are too specific for the initial implementation. I don't understand what you mean by "uniquely identify an invocation".

>  I agree that this can be tracked by individually by the hooks, but it is part of the identity of an invocation of a callback. (i.e., an execution of a callback is uniquely identifed by the tuple [asyncId, invocationId])

The point of this API at this moment is to offer only that which can't be gained by any other means (with exception to `.scope()`, but added to prevent easily hazardous scenarios). I'm not sure what you mean by `invocationId`. If you simply mean a unique id for every call to `MakeCallback()` that would be the simplest piece of information to track in user land. A simple `invocationId++;` would do the trick.

Though I think what you may be asking for is assigning an `invocationId` in the `init()` area so it can be tracked and identified in the `pre()`. Is this correct?